src/storage/txn/tracker.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2025 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cell::RefCell;
+
+use tikv_util::time::Instant;
+use tracker::{FutureTrack, TrackerToken, GLOBAL_TRACKERS};
+
+use crate::storage::metrics::CommandKind;
+
+thread_local! {
+    // The TLS state exists solely for the purpose of [`TlsFutureTracker::collect_to_tracker`]
+    // Remove this complication if we can find a better way to collect the time.
+    static CURRENT_STATE: RefCell<Option<State>> = const { RefCell::new(None) };
+}
+
+#[derive(Debug, Clone, Copy)]
+enum PollState {
+    None,
+    Begin(Instant),
+    Finish(Instant),
+}
+
+#[derive(Debug)]
+struct State {
+    token: TrackerToken,
+    tag: CommandKind,
+    cid: u64,
+    current_stage: PollState,
+
+    future_process_nanos: u64,
+    future_suspend_nanos: u64,
+}
+
+pub struct TlsFutureTracker {
+    state: Option<State>,
+}
+
+impl TlsFutureTracker {
+    pub fn new(token: TrackerToken, tag: CommandKind, cid: u64) -> Self {
+        let state = State {
+            token,
+            tag,
+            cid,
+            // Set to Finish state to tracking the wait time before the first poll.
+            current_stage: PollState::Finish(Instant::now()),
+
+            future_process_nanos: 0,
+            future_suspend_nanos: 0,
+        };
+        TlsFutureTracker { state: Some(state) }
+    }
+
+    /// Collects the current future process and suspend time to the tracker.
+    ///
+    /// It exists because a storage callback is invoked during the future
+    /// polling. To make measurements more accurate, we need to collect the
+    /// time before the callback is invoked.
+    pub fn collect_to_tracker(now: Instant, tracker: &mut tracker::Tracker) {
```

Comment:
Perhaps it's also needed to be called in the future task of `fail_fast_or_check_deadline`, before the `early_response` and callback executed after timeout.